### PR TITLE
Add decorators version `2023-05`

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The new version implements 1 (commit: https://github.com/babel/babel/pull/15570/commits/d222bb5a7044516c299be19e49c9190791e669f1) and 6 (commit: https://github.com/babel/babel/pull/15570/commits/cc101de4deb4938a7c666e09a857fe9bcbb034bf) from https://github.com/tc39/proposal-decorators/issues/499. All the other points were already behaving like that in previous versions of the transform.

cc @pzuraq 

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15570"></a>

### CODE REVIEW

File: packages/babel-helpers/src/helpers/applyDecs2303.js [hunk 1]
```diff
@@ -537,7 +519,7 @@ function applyClassDecs(targetClass, classDecs) {
     [
       // member decorators
       [
-        dec,                // dec or array of decs
+        decs,               // array of decs or of decs and this values
```
Comment (JLHwung) [inline packages/babel-helpers/src/helpers/applyDecs2303.js:522]: Are they always "array of decorator and this values" now?

**Review by JLHwung — APPROVED** [review]
lgtm, left a comment about flags packing.

File: packages/babel-helpers/src/helpers/applyDecs2303.js [hunk 2]
```diff
@@ -494,18 +469,25 @@ function applyClassDecs(targetClass, classDecs) {
     var newClass = targetClass;
     var name = targetClass.name;
 
-    for (var i = classDecs.length - 1; i >= 0; i--) {
+    var hasThis = classDecs[0] === 1;
```
[the remark this answers is not in the document]
Comment (nicolo-ribaudo) [reply packages/babel-helpers/src/helpers/applyDecs2303.js]: Done!

File: packages/babel-helpers/src/helpers/applyDecs2303.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-helpers/src/helpers/applyDecs2303.js]: Now that the flag is stored together with the other flags, we can optimize the array away again when we have a single decorator without `this` :)

**Comment by nicolo-ribaudo** [comment]
Diff ignoring the `chore` commits: https://github.com/babel/babel/pull/15570/files/a93f50195594f664f585f1f3d759a4800fa2bae4..9ed66d594cf8d178069c7a8418fe83f9d84131d5

**Review by JLHwung — APPROVED** [review]

**Comment by JLHwung** [comment]
Added a new test case per https://github.com/babel/babel/issues/15608#issuecomment-1533647975.

File: packages/babel-helpers/src/helpers/applyDecs2301.js [hunk 3]
```diff
@@ -300,392 +274,437 @@ function applyMemberDec(
...
-          addInitializer: createAddInitializerMethod(
-            initializers,
-            decoratorFinishedRef
-          ),
-        });
-      } finally {
-        decoratorFinishedRef.v = true;
-      }
+  function applyClassDecs(targetClass, classDecs) {
+    if (classDecs.length > 0) {
```
Comment (liuxingbaoyu) [inline packages/babel-helpers/src/helpers/applyDecs2301.js:497]:
```suggestion
    if (classDecs.length) {
```

**Review by liuxingbaoyu — APPROVED** [review]
Sorry I'm not familiar with the decorators proposal, just looked at some code styles.

File: packages/babel-helpers/src/helpers/applyDecs2301.js [hunk 4]
```diff
@@ -300,392 +274,437 @@ function applyMemberDec(
...
+        init = function (instance, init) {
+          var value = init;
 
-        return value;
-      };
-    } else {
-      var originalInitializer = init;
+          for (var i = 0; i < ownInitializers.length; i++) {
+            value = ownInitializers[i].call(instance, value);
+          }
```
Comment (liuxingbaoyu) [inline packages/babel-helpers/src/helpers/applyDecs2301.js:349]: Can we use `for of` in helpers? I'm not sure if it's allowed in helpers.

Comment (nicolo-ribaudo) [reply packages/babel-helpers/src/helpers/applyDecs2301.js:349]: We try to keep helpers es5-only _when possible_ (or es5 syntax + polyfillable functions)

**Comment by nicolo-ribaudo** [comment]
Yes! I will try to get to it before the end of the week.

**Comment by nicolo-ribaudo** [comment]
@pzuraq Could you review [`1bbecfa` (#15570)](https://github.com/babel/babel/pull/15570/commits/1bbecfaf31963982bee07494e0fbce840bd5912e)?

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-05-ordering/accessor-init-and-set-consistent/exec.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    init(v) {
+      return v * 4;
+    },
+  };
+}
+
+class Foo {
+  @minusTwo @timesFour accessor bar = 5;
+}
+debugger;
```
Comment (pzuraq) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2023-05-ordering/accessor-init-and-set-consistent/exec.js:26]: should this be here?

**Review by pzuraq — APPROVED** [review]
Looks good! Just one small thing (not related to implementation)

File: packages/babel-helpers/src/helpers/applyDecs2305.js [hunk 6]
```diff
@@ -319,7 +319,7 @@ function applyMemberDec(
       init = function (instance, init) {
         var value = init;
 
-        for (var i = 0; i < ownInitializers.length; i++) {
+        for (var i = ownInitializers.length - 1; i >= 0; i--) {
```
Comment (JLHwung) [inline packages/babel-helpers/src/helpers/applyDecs2305.js:322]: If a decorator calls `context.addInitializer()` twice:
```js
context.addInitializer(f);
context.addInitializer(g);
```
Previously `f` will be invoked before `g` when the class member is initialized, now `f` will be invoked _after_ `g`. @pzuraq is that right?

Comment (pzuraq) [reply packages/babel-helpers/src/helpers/applyDecs2305.js:322]: No, the order of `addInitializer` initializers has not been changed, only initializers returned from `init`

**Pull request merged**